File: 695-max-area-of-island/max-area-of-island.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var maxAreaOfIsland = function(grid) {
    if (!grid || grid.length === 0) return 0

    let maxArea = 0
    const directions = [[1,0],[-1,0],[0,1],[0,-1]]

    for (let r = 0; r < grid.length; r++) {
        for (let c = 0; c < grid[0].length; c++) {
            if (grid[r][c] === 1) {
                const currentArea = dfs(grid, directions, r, c)
                maxArea = Math.max(maxArea, currentArea)
            }
        }
    }


    return maxArea
};

function dfs(grid, directions, r, c) {
    if (r < 0 || c < 0 || r >= grid.length || c >= grid[0].length || grid[r][c] === 0) {
        return 0
    }
    grid[r][c] = 0
    let area = 1

    for (const [dr, dc] of directions) {
        const newR = r + dr
        const newC = c + dc
        area += dfs(grid, directions, newR, newC)

    }

    return area
}

File: 695-max-area-of-island/max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        rows, cols = len(grid), len(grid[0])
        directions = [(-1,0),(1,0),(0,-1),(0,1)]
        visited = set()

        def dfs(r, c):
            if r < 0 or r >= rows or c < 0 or c >= cols or (r,c) in visited or grid[r][c] == 0:
                return 0

            visited.add((r,c))
            area = 1

            for dr, dc in directions:
                area += dfs(r + dr, c + dc)

            return area
        
        max_area = 0

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 1:
                    max_area = max(max_area, dfs(r, c))

        return max_area